backend/danswer/connectors/notion/connector.py

```diff
@@ -0,0 +1,222 @@
+class NotionConnector(LoadConnector, PollConnector):
+    """Notion Page connector that reads all Notion pages
+    this integration has been granted access to.
+
+    Arguments:
+        batch_size (int): Number of objects to index in a batch
+    """
+
+    def __init__(self, batch_size: int = INDEX_BATCH_SIZE) -> None:
+        """Initialize with parameters."""
+        self.batch_size = batch_size
+        self.headers = {
+            "Content-Type": "application/json",
+            "Notion-Version": "2022-06-28",
+        }
+
+    def load_credentials(self, credentials: dict[str, Any]) -> dict[str, Any] | None:
+        """Applies integration token to headers"""
+        self.headers[
+            "Authorization"
+        ] = f'Bearer {credentials["notion_integration_token"]}'
+        return None
+
+    def _read_blocks(self, block_id: str, num_tabs: int = 0) -> str:
+        """Reads blocks for a page"""
+        done = False
+        result_lines_arr = []
+        cur_block_id = block_id
+        while not done:
+            block_url = f"https://api.notion.com/v1/blocks/{cur_block_id}/children"
+            query_dict: Dict[str, Any] = {}
+
+            res = requests.request(
+                "GET", block_url, headers=self.headers, json=query_dict
+            )
+            data = res.json()
+
+            for result in data["results"]:
+                result_type = result["type"]
+                result_obj = result[result_type]
+
+                cur_result_text_arr = []
+                if "rich_text" in result_obj:
+                    for rich_text in result_obj["rich_text"]:
+                        # skip if doesn't have text object
+                        if "text" in rich_text:
+                            text = rich_text["text"]["content"]
+                            prefix = "\t" * num_tabs
+                            cur_result_text_arr.append(prefix + text)
+
+                result_block_id = result["id"]
+                has_children = result["has_children"]
+                if has_children:
+                    children_text = self._read_blocks(
+                        result_block_id, num_tabs=num_tabs + 1
+                    )
+                    cur_result_text_arr.append(children_text)
+
+                cur_result_text = "\n".join(cur_result_text_arr)
+                result_lines_arr.append(cur_result_text)
+
+            if data["next_cursor"] is None:
+                done = True
+                break
+            else:
+                cur_block_id = data["next_cursor"]
+
+        result_lines = "\n".join(result_lines_arr)
+        return result_lines
+
+    def _read_pages(self, pages: List[NotionPage]) -> List[Document]:
+        """Reads pages for rich text content and generates Documents"""
+        docs_batch = []
+        for page in pages:
+            page_text = self._read_blocks(page.id)
+            page_title = page.properties.get("Name", None) or page.properties.get(
```

Comment:
![image](https://github.com/danswer-ai/danswer/assets/32520769/3854394d-f4d4-41ac-bfd9-67018bfa3bce)

Seems to be some slight bug here with fetching the page title in some cases.